Add push subscriptions to apsd ignore list if necessary

bnham opened the pull request:
#### 94e15b227a3ceda11fadfb31f3bde8e9d40d5ae8
<pre>
Add push subscriptions to apsd ignore list if necessary
<a href="https://bugs.webkit.org/show_bug.cgi?id=242719">https://bugs.webkit.org/show_bug.cgi?id=242719</a>
&lt;rdar://problem/91586206&gt;

Reviewed by Chris Dumez.

Currently, if a user changes the notification permission for an origin from Allow to Deny in the
browser settings pane, we don&apos;t communicate this change to apsd. This means that pushes from this
origin can continue to wake the device and induce battery drain even though the push will always be
useless, as we bail out of running push event handlers early when we detect that an origin is
missing the notification permission.

To fix this, we now add or remove push topics from the apsd ignore list when the user denies or
allows push permissions for an origin (in the call to providerDidUpdateNotificationPolicy).

Note that moving a topic to the ignore list does not delete the associated PushSubscription. It&apos;s
more like putting the subscription in a paused state that can be re-enabled later by adding the
topic back to the enabled list. This matches the behavior of native app push subscriptions when the
notification preference is toggled for native apps.

* Source/WebCore/Modules/push-api/PushDatabase.cpp: Update schema to add a state column that tracks enabled/ignored state.
(WebCore::PushTopics::isolatedCopy const):
(WebCore::PushTopics::isolatedCopy):
(WebCore::PushDatabase::insertRecord):
(WebCore::PushDatabase::getTopics):
(WebCore::PushDatabase::setPushesEnabledForOrigin):
* Source/WebCore/Modules/push-api/PushDatabase.h:

* Source/WebKit/NetworkProcess/NetworkProcess.cpp: Forward the setPushAndNotificationsEnabledForOrigin message to webpushd.
(WebKit::NetworkProcess::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::messageTypeSendsReply):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp: Call setPushAndNotificationsEnabledForOrigin when the user changes the notification permissions from Allow to Deny or vice-versa.
(WebKit::persistentDataStoreIfExists):
(WebKit::setPushesAndNotificationsEnabledForOrigin):
(WebKit::removePushSubscriptionsForOrigins):
(WebKit::WebNotificationManagerProxy::providerDidUpdateNotificationPolicy):
* Source/WebKit/webpushd/MockPushServiceConnection.h:
* Source/WebKit/webpushd/MockPushServiceConnection.mm:
* Source/WebKit/webpushd/PushService.h:
(WebPushD::PushService::enabledTopics):
(WebPushD::PushService::ignoredTopics):

* Source/WebKit/webpushd/PushService.mm: Move topics to the enabled or ignored list as necessary when the database state changes.
(WebPushD::updateTopicLists):
(WebPushD::PushService::setPushesEnabledForBundleIdentifierAndOrigin):

* Source/WebKit/webpushd/WebPushDaemon.h: Add a new raw XPC message type that can be used for sending messages from integration tests directly to webpushd without going through NetworkProcess.
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::connectionEventHandler):
(WebPushD::Daemon::decodeAndHandleRawXPCMessage):
(WebPushD::Daemon::decodeAndHandleMessage):
(WebPushD::Daemon::setPushAndNotificationsEnabledForOrigin):
(WebPushD::toXPCArray):
(WebPushD::Daemon::getPushTopicsForTesting):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::getTopicsSync):
(TestWebKitAPI::PushDatabaseTest::getTopics):
(TestWebKitAPI::PushDatabaseTest::setPushesEnabledForOrigin):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::addMessageTypeHeaders):
(TestWebKitAPI::createMessageDictionary):
(TestWebKitAPI::toStringVector):

Canonical link: <a href="https://commits.webkit.org/252434@main">https://commits.webkit.org/252434@main</a>
</pre>

cdumez commented on Source/WebCore/Modules/push-api/PushDatabase.cpp:
```diff
@@ -624,21 +639,34 @@ void PushDatabase::getIdentifiers(CompletionHandler<void(HashSet<PushSubscriptio
+        }
 
-        completeOnMainQueue(WTFMove(completionHandler), WTFMove(topics));
+        completeOnMainQueue(WTFMove(completionHandler), topics);
```
Why are we dropping the WTFMove here? Doesn't it mean we're losing out of the optimized isolatedCopy() on a r-value?

bnham replied:
Yeah I accidentally dropped this while refactoring and will add it back in.

cdumez commented on Source/WebKit/webpushd/PushService.h:
```diff
@@ -57,10 +57,15 @@ class PushService {
     PushServiceConnection& connection() { return m_connection; }
     WebCore::PushDatabase& database() { return m_database; }
 
+    Vector<String> enabledTopics() { return m_connection->enabledTopics(); }
```
Ideally this function would be const and return a `const Vector<String>`

bnham replied:
This eventually bottoms out in `PushServiceConnection::enabledTopics()`, which has to return a `Vector<String>` rather than `const Vector<String>&`. This is because `ApplePushServiceConnection::enabledTopics()` (which is another subclass in this class hierarchy and the one that actually matters in the real implementation) has to return `Vector<String>` since it doesn't cache the topics locally in a member variable. Instead it always constructs a `Vector<String>` on the fly by converting it from the `NSArray<NSString *>` given to it from APS. So there is always at least one copy involved.

Given this, making the return type `Vector<String>` and counting on move semantics to move the copy up the stack seems okay to me.

cdumez commented on Source/WebKit/webpushd/PushService.h:
```diff
@@ -57,10 +57,15 @@ class PushService {
     WebCore::PushDatabase& database() { return m_database; }
 
+    Vector<String> enabledTopics() { return m_connection->enabledTopics(); }
+    Vector<String> ignoredTopics() { return m_connection->ignoredTopics(); }
```
ditto.

bnham replied:
Same comment as above.

cdumez commented on Source/WebKit/webpushd/MockPushServiceConnection.h:
```diff
@@ -39,19 +39,25 @@ class MockPushServiceConnection final : public PushServiceConnection {
-    Vector<String> ignoredTopics() override;
-    Vector<String> opportunisticTopics() override;
-    Vector<String> nonWakingTopics() override;
+    Vector<String> enabledTopics() override { return m_enabledTopics; }
```
Can this be const and return a `const Vector<String>`?

bnham replied:
Same comment as above.

cdumez commented on Source/WebKit/webpushd/PushService.mm:
```diff
@@ -521,6 +523,23 @@ void reject(WebCore::ExceptionData&& data)
+        if (!recordsChanged)
+            return handler();
+
+        updateTopicLists(m_connection, m_database, [handler = WTFMove(handler)]() mutable {
```
Wouldn't this work?
`updateTopicLists(m_connection, m_database, WTFMove(handler));`

bnham replied:
Yep, will fix this.

bnham committed:
Add push subscriptions to apsd ignore list if necessary
https://bugs.webkit.org/show_bug.cgi?id=242719
<rdar://problem/91586206>

Reviewed by Chris Dumez.

Currently, if a user changes the notification permission for an origin from Allow to Deny in the
browser settings pane, we don't communicate this change to apsd. This means that pushes from this
origin can continue to wake the device and induce battery drain even though the push will always be
useless, as we bail out of running push event handlers early when we detect that an origin is
missing the notification permission.

To fix this, we now add or remove push topics from the apsd ignore list when the user denies or
allows push permissions for an origin (in the call to providerDidUpdateNotificationPolicy).

Note that moving a topic to the ignore list does not delete the associated PushSubscription. It's
more like putting the subscription in a paused state that can be re-enabled later by adding the
topic back to the enabled list. This matches the behavior of native app push subscriptions when the
notification preference is toggled for native apps.

* Source/WebCore/Modules/push-api/PushDatabase.cpp: Update schema to add a state column that tracks enabled/ignored state.
(WebCore::PushTopics::isolatedCopy const):
(WebCore::PushTopics::isolatedCopy):
(WebCore::PushDatabase::insertRecord):
(WebCore::PushDatabase::getTopics):
(WebCore::PushDatabase::setPushesEnabledForOrigin):
* Source/WebCore/Modules/push-api/PushDatabase.h:

* Source/WebKit/NetworkProcess/NetworkProcess.cpp: Forward the setPushAndNotificationsEnabledForOrigin message to webpushd.
(WebKit::NetworkProcess::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.cpp:
(WebKit::NetworkNotificationManager::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/NetworkProcess/Notifications/NetworkNotificationManager.h:
* Source/WebKit/Shared/WebPushDaemonConstants.h:
(WebKit::WebPushD::messageTypeSendsReply):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setPushAndNotificationsEnabledForOrigin):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:

* Source/WebKit/UIProcess/Notifications/WebNotificationManagerProxy.cpp: Call setPushAndNotificationsEnabledForOrigin when the user changes the notification permissions from Allow to Deny or vice-versa.
(WebKit::persistentDataStoreIfExists):
(WebKit::setPushesAndNotificationsEnabledForOrigin):
(WebKit::removePushSubscriptionsForOrigins):
(WebKit::WebNotificationManagerProxy::providerDidUpdateNotificationPolicy):
* Source/WebKit/webpushd/MockPushServiceConnection.h:
* Source/WebKit/webpushd/MockPushServiceConnection.mm:
* Source/WebKit/webpushd/PushService.h:
(WebPushD::PushService::enabledTopics):
(WebPushD::PushService::ignoredTopics):

* Source/WebKit/webpushd/PushService.mm: Move topics to the enabled or ignored list as necessary when the database state changes.
(WebPushD::updateTopicLists):
(WebPushD::PushService::setPushesEnabledForBundleIdentifierAndOrigin):

* Source/WebKit/webpushd/WebPushDaemon.h: Add a new raw XPC message type that can be used for sending messages from integration tests directly to webpushd without going through NetworkProcess.
* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::Daemon::connectionEventHandler):
(WebPushD::Daemon::decodeAndHandleRawXPCMessage):
(WebPushD::Daemon::decodeAndHandleMessage):
(WebPushD::Daemon::setPushAndNotificationsEnabledForOrigin):
(WebPushD::toXPCArray):
(WebPushD::Daemon::getPushTopicsForTesting):
* Tools/TestWebKitAPI/Tests/WebCore/PushDatabase.cpp:
(TestWebKitAPI::getTopicsSync):
(TestWebKitAPI::PushDatabaseTest::getTopics):
(TestWebKitAPI::PushDatabaseTest::setPushesEnabledForOrigin):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebPushDaemon.mm:
(TestWebKitAPI::addMessageTypeHeaders):
(TestWebKitAPI::createMessageDictionary):
(TestWebKitAPI::toStringVector):

Canonical link: https://commits.webkit.org/252434@main
